### DISCUSSION TITLE
Trim down the size of the compiled binary

### DISCUSSION THREAD

**Discussion by laolarou726** [AvaloniaUI/Avalonia#9217] [Ideas]
Are there any ways to reduce the compiled binary size? For now, the compiled binary size is around 60~80 MB. Which is really large. Are there any plans in the future version to consider to reduce the binary size of the Avalonia build?

**Comment by maxkatz6** [comment]
Avalonia assemblies don't add too much of size (except of Fluent theme because it contains fonts). SkiaSharp though eats almost 10mb unzipped.
Most of bundle size is taken by dotnet runtime itself, which isn't famous by small assembly sizes. There are ways to improve it though well documented by Microsoft. (see https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained, https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-6-0, https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli)

Baseline:
```
dotnet new avalonia.app
dotnet publish -c Release -r win-x64 -p:PublishTrimmed=true
```
Produces 58.2 MB unzipped and 25 MB zipped for the minimal app. With more XAML views it will grow a little. Majority of additional size will be added by third party dependencies.

Next, if I make it a single file:
```
dotnet publish -c Release -r win-x64 -p:PublishTrimmed=true -p:PublishSingleFile=true
```
50.8 MB unzipped, 22 MB zipped.

Next, if I allow "link" trimming:
```
dotnet publish -c Release -r win-x64 -p:PublishTrimmed=true -p:PublishSingleFile=true -p:TrimMode=link
```
37.4 MB unzipped, 16.1 MB zipped.
Note, that I also had to add TrimmerRootDescriptor file, something [like here](https://github.com/AvaloniaUI/Avalonia/issues/9127#issuecomment-1272443640). 

Last step is something that should be improved on Avalonia side, so developers won't need to apply additional configuration. Also, I can't confirm if it perfectly works with Avalonia 0.10 (test app is running though), more work was done for Avalonia 11.0.

**Reply by maxkatz6** [reply]
Note, if you want to speed up startup time of your app by using [PublishReadyToRun](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run) (which we do recommend), it will increase resulting time even more. But that's a tradeoff of startup time and app size, which can be controlled by [PublishReadyToRunExclude](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run#how-is-the-set-of-precompiled-assemblies-chosen).

**Reply by laolarou726** [reply]
Will the AOT become one of the official build option for Avalonia?

**Reply by laolarou726** [reply]
Maybe get rid of all the reflections in the Avalonia source is a great start?

**Reply by maxkatz6** [reply]
> Maybe get rid of all the reflections in the Avalonia source is a great start

Yes, and it was mostly done already. 

> Will the AOT become one of the official build option for Avalonia?

That's an option for .NET, not for Avalonia. Remember, that size, trimming and startup optimizations is a job for the runtime and SDK, not for the Avalonia. 
But Avalonia needs to be compatible with these things, what we are trying to do.

**Reply by maxkatz6** [reply]
Also, Avalonia works with AOT, you can find discussions/issues about it in this repo

**Reply by robloo** [reply]
I didn't realize Fluent theme had fonts. Why is that?

**Reply by maxkatz6** [reply]
To not depend on system fonts and to have same fonts for the one theme.
https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Themes.Fluent/Assets

**Reply by robloo** [reply]
That should be done at app level... or with another fonts nuget package IMO. Very strange to package fonts up with a theme like this based on other XAML frameworks. It's expected to use the system fonts.

**Reply by yingDev** [reply]
> To not depend on system fonts and to have same fonts for the one theme. https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Themes.Fluent/Assets

Maybe these fonts are redundant because they won't work with other non-English environments.

**Comment by DamianSuess** [comment]
As a tip to reduce the size even more, developers can employ a couple of more options. This is most helpful when debugging with tools such as [VS Linux Debugger](https://github.com/SuessLabs/VsLinuxDebug) where the payload deployed is your `bin` folder.

1. Specify the languages used. This reduces Microsoft.CodeAnalysis' language files/folders (can reduce up to 54 files)
    * https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#satelliteresourcelanguages
2. Set the Runtimes supported
    * [MSBuild Property - RuntimeIdentifiers](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#runtimeidentifiers)
    * [.NET Runtime Identifier Catalog](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog)

To apply, update your `.csproj` and add the following lines. This is supposed to drop our install size significantly.

```xml
  <PropertyGroup>
    <!-- Reduce Build Resources -->
    <!-- Supported languages https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#satelliteresourcelanguages -->
    <SatelliteResourceLanguages>en</SatelliteResourceLanguages>
    <!-- Supported runtimes (https://learn.microsoft.com/en-us/dotnet/core/rid-catalog) -->
    <RuntimeIdentifiers>win10-x64;linux-x64</RuntimeIdentifiers>
  </PropertyGroup>
```

UPDATE:
* I opened a bug with https://github.com/mono/SkiaSharp/issues/2520 as it is still generating the runtime files.

**Reply by maxkatz6** [reply]
In case of runtime identifiers, it's worth of defining only a single runtime while you are publishing final app. 
With Avalonia in mind, it will trim linux backend from win version and vice versa.

**Reply by DamianSuess** [reply]
Thanks, @maxkatz6. In our use case, our applications are targeting Windows and Linux runtimes while attempting to use a single project. Perhaps, our implementation is incorrect?

p.s. i updated the original post to include a link to the SkiaSharp issue

**Reply by maxkatz6** [reply]
@DamianSuess I was primarily meaning the use case, when developers do different app builds per each platform, when you can specify only one target RID. This way it can be trimmed per specific platform. If developers decide to keep one universal build with multiple RIDs, there is no problem with that.

**Reply by DamianSuess** [reply]
@maxkatz6, thanks for the clarification. Sorry, if I muddied the waters in this thread.

Overall, this journey to minify the debug/release output size of our builds has led me down the rabbit hole to try and figure out if there is a bug with the `<RuntimeIdentifiers>` (plural) or if I'm using it incorrectly. So far, I'm leaning toward, a bug.  If so, is the issue on the _Avalonia.SkiaSharp_, SkiaSharp, or [DotNet Runtime](https://github.com/dotnet/runtime) side? 🤔   Gotta love software dev.

Until this gets pinned down, I'll stick with post-build deleting of files.

**Comment by Numpsy** [comment]
Hi,

I think this is rather trivial in terms of overall size, but just in case:
I was just looking at a Windows-x64 / single file pub;ished / trimmed / net7.0 binary in ILSpy trying to see if there's anything more that can be trimmed from it, and I notice that there are some references to Avalonia.Metal and Metal APIs present:

![image](https://github.com/AvaloniaUI/Avalonia/assets/1178570/a56fd98a-9b1d-41d5-8dd4-7625f1393266)

So I was just wondering if it should be possible for these to be trimmed out on a Windows only build?

**Reply by maxkatz6** [reply]
It might be trimmed if you publish your app as trimmed and for win-x64/amd RID. But either way, it's a very minor dependency, as it just a thin abstraction layer between avalonia and platforms that might support metal.

**Comment by tkefauver** [comment]
From @maxkatz6 helpful info if you can't publish from command line (like with msix, etc.) you can add this to your target`.csproj` file 

```
<PublishTrimmed>true</PublishTrimmed>
<PublishSingleFile>true</PublishSingleFile>
<TrimMode>link</TrimMode>
```

Reduced my installer from 45mb to 15mb! 😎

**Reply by sgf** [reply]
Good job but others may only need 1-2mb and only extract 5mb. 😂
